Create script to assign osg GIDs to OSPool projects (SOFTWARE-5619)

williamnswanson opened the pull request:
Initial Commit for project osggid assignment script

williamnswanson committed:
Initial Commit for project osggid assignment script

williamnswanson committed:
whitespace change

Mis-config of the Black code formatter caused some unhelpful attempts at saving line length.

williamnswanson committed:
Move ospool pattern and initial project GID

For quicker possible future changes and creating command line options. Also slight clean up.

jasoncpatton commented on group_identifier_assigner.py:
```diff
@@ -0,0 +1,262 @@
+    found_list = list((id["Type"] == id_type for id in id_list))
+    if any(found_list):
+        return found_list.index(True)
+    return -1
```
This works, but I think you can use a list comprehension directly for `found_list` (`[id["Type"] == id_type for id in id_list]`) rather than creating and consuming a generator, but even better, don't bother with any comparisons except for the `index` lookup using a `try`/`except` to retain the same return value behavior.
```suggestion
def identifier_index(id_list, id_type):
    id_type_list = [id["Type"] for id in id_list]
    try:
        return id_type_list.index(id_type)
    except ValueError:
        return -1
```

williamnswanson replied:
I've incorporated this into [575a016](https://github.com/opensciencegrid/osg-comanage-project-usermap/pull/18/commits/575a016d902393d252c002fd71958ebaa3be4cdc)

jasoncpatton commented on group_identifier_assigner.py:
```diff
@@ -0,0 +1,262 @@
+POST = "POST"
+DELETE = "DELETE"
+
+OSPOOL_PATTERN_STR = "Yes-"
```
Nitpick: this is not so much a "pattern" but a "prefix".

williamnswanson replied:
Gotcha', changed in commit [575a016](https://github.com/opensciencegrid/osg-comanage-project-usermap/pull/18/commits/575a016d902393d252c002fd71958ebaa3be4cdc)

jasoncpatton commented on group_identifier_assigner.py:
```diff
@@ -0,0 +1,262 @@
+            else:
+                is_project = False
+
+            osggid_index = identifier_index(identifier_list, "osggid")
+            if osggid_index != -1:
+                highest_osggid = max(highest_osggid, int(identifier_list[osggid_index]["Identifier"]))
+            elif is_project is True:
+                # for each, set a 'OSG GID' starting from 200000 and a 'OSG Group Name' that is the group name
+                osggid_to_assign = max(highest_osggid + 1, options.project_gid_startval)
+                highest_osggid = osggid_to_assign
+                add_identifier_to_group(gid, type="osggid", identifier_name=osggid_to_assign)
```
I see the logic here in that `highest_osggid` is never able to decrement, but I think there is still a possibility for collision unless `co_groups.json` is already somehow sorted by `osggid` (but happy to be proven wrong!). For instance, suppose we already have assignments (or missing assignments):
```
ProjectA -> 200000
ProjectB -> 200001
ProjectC -> (undefined)
ProjectD -> 200002
```
and `co_groups` happens to go in order of `ProjectA, ProjectB, ProjectC, ProjectD`, I believe that this script will try to assign ProjectC a `osggid` of `200002`.

I think to be certain, the script should go through all the project groups once to fetch the highest existing `osggid` (if any) and only then start assigning new GIDs.

williamnswanson replied:
Ok, I've moved the assigning out to a separate loop at the end in [575a016](https://github.com/opensciencegrid/osg-comanage-project-usermap/pull/18/commits/575a016d902393d252c002fd71958ebaa3be4cdc)

jasoncpatton reviewed (changes requested):
See comments, main concern is that there is a possibility for OSG GID collision.

williamnswanson committed:
Updates from Jason review

Also "replace"-ed repoveprefix, since I think this script might be running on a machine with only python 3.6.8 (removeprefix needs 3.9)

jasoncpatton commented on group_identifier_assigner.py:
```diff
@@ -0,0 +1,266 @@
+                highest_osggid = max(highest_osggid, int(identifier_list[osggid_index]["Identifier"]))
+            elif is_project is True:
+                project_name = project_id.replace(OSPOOL_PROJECT_PREFIX_STR, "", 1).lower()
+                projects_to_assign_identifiers.append(tuple([gid, project_name]))
```
You can tack on an extra comma to let Python know you want a 2-tuple using parentheses (i.e. without converting a list):
```suggestion
                projects_to_assign_identifiers.append((gid, project_name,))
```

jasoncpatton reviewed (changes requested):
One small change, otherwise LGTM!

williamnswanson committed:
Update group_identifier_assigner.py

Co-authored-by: Jason Patton <jason.c.patton@gmail.com>
